### PR TITLE
Add distinguished select game screen

### LINKED ISSUE #2173: Distinguish the "Select Game" screen for single player vs host game
At present the "Select Game" screen is identical whether you're hosting a server or playing single player. We should probably have a sub-heading or something to indicate which it is.

Easy!

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/31106707-distinguish-the-select-game-screen-for-single-player-vs-host-game?utm_campaign=plugin&utm_content=tracker%2F5279&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F5279&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by BdeBock**
fixes #2173 
Add distinguished select game screen for singleplayer and host game.
Singleplayer now has a subtitle "singleplayer" and host game has a subtitle "multiplayer".
Created a SelectGameScreenHost class (copy of SelectGameScreen) just to be able to couple the right .ui file to it (no idea if this is the right way to make the distinction)

### CODE REVIEW

**Comment by GooeyHub** [comment]
Can one of the admins please verify this patch?

**Comment by msteiger** [comment]
Hi Bas - what's wrong?

**Comment by BdeBock** [comment]
Hi Martin, I fixed the issue, but wanted my team ( @AlexanderGrooff , @JanZegers , @arkka ) to go over it first before pull requesting it to you guys. Though as you can see I accidentally created the PR already, that's why I closed it again.

**Comment by GooeyHub** [comment]
Can one of the admins please verify this patch?

**Comment by Cervator** [comment]
ok to test

Hi @BdeBock and thanks for the PR :-)

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/440/
Hooray Jenkins reported success with all tests good!

**Comment by msteiger** [comment]
Thanks for your contribution, Bas. I'd prefer to use one class for both cases, since they are highly similar, otherwise this results in [code duplication](https://en.wikipedia.org/wiki/Duplicate_code#Problems_associated_with_duplicate_code). You can still set different subtitles, if that's the only difference between the two.

In fact, the class already differentiates between the two modes in the action listener for the "Start Game" button (the `isLoadingAsServer` flag controls the behavior).

You should be able to rename/reuse the flag to respect the subtitle id as well.

**Comment by AlexanderGrooff** [comment]
It seems that the SelectGameScreenHost.java was never reached, meaning that it's effectively dead code.

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/442/
Hooray Jenkins reported success with all tests good!

**Comment by msteiger** [comment]
Thanks!
The `selectGameScreenHost` UI code looks like a duplicate as well.
Also, `mainmenu.skin/subtitle` could be redundant as it is probably defined in `default.skin` already.

**Comment by Cervator** [comment]
Bump - should we merge this or are we still waiting for a caption tweak instead of unique new `.ui` file?

**Comment by msteiger** [comment]
I think that only 5-6 lines of code need to be changed so solve the issue, so I'd either wait until this PR is updated or create a new fix.

**Comment by BdeBock** [comment]
Kind of busy at the moment, will fix it in the weekend.

**Comment by Cervator** [comment]
No problem, just wanted to check :-)

**Comment by BdeBock** [comment]
Hi guys, looked at it again. 
Martin mentioned selectGameScreenHost.ui looks a lot like selectGameScreen.ui. This is true except for the difference of the subtitle label. But I see no other way to let the selectGameScreen of singleplayer and multiplayer have a different sub-heading, without letting them have their own .ui file. 
For the second point Martin mentioned, the mainmenu.skin/subtitle is defined as there wasn't one yet for a subtitle/sub-heading. And I believe it should be placed in the mainmenu.skin as it is part of the mainmenu. 

Could be I'm totally wrong about this, if so please let me know ;p

**Comment by Cervator** [comment]
Hey again @BdeBock thanks for following up :-)

I figure there's some way to use a variable in the .ui file that in turn will point to an i18n string key. I just don't know how.

Hopefully @msteiger will see this tomorrow and know of an answer :-)

**Comment by msteiger** [comment]
You can use a placeholder text for the subtitle label and set the correct one through Java code when the menu screen is opened.

This is done at several occasions, for example in the module selection dialog, which updates the module description text according to the module selection.

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/479/
Hooray Jenkins reported success with all tests good!

**Comment by BdeBock** [comment]
Implemented the proposed solution of msteiger, think everything is ok now.

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/480/
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/481/
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/MainMenuScreen.java [hunk 1]
```diff
@@ -34,7 +34,12 @@
     @Override
     public void initialise() {
         find("version", UILabel.class).setText(TerasologyVersion.getInstance().getHumanVersion());
-        WidgetUtil.trySubscribe(this, "singleplayer", button -> getManager().pushScreen("engine:selectGameScreen"));
+        WidgetUtil.trySubscribe(this, "singleplayer", button -> {
+            UIScreenLayer screen = getManager().pushScreen("engine:selectGameScreen");
```
Comment (msteiger) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/MainMenuScreen.java]: There should be a similar method with second parameter that takes a class argument. It returns an arbitrary type which makes the check below unnecessary.

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/SelectGameScreen.java [hunk 2]
```diff
@@ -47,6 +49,20 @@
 
     @Override
     public void initialise() {
+
+        UILabel subtitle2 = find("subtitle2", UILabel.class);
```
Comment (msteiger) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/SelectGameScreen.java]: Good! Can you find a better name than `subtitle2`?

File: engine/src/main/resources/assets/skins/mainmenu.skin [hunk 3]
```diff
@@ -35,6 +35,10 @@
             "font": "NotoSans-Regular-Large",
             "text-color": "F0F0F0FF"
         },
+         "subtitle": {
```
Comment (msteiger) [inline engine/src/main/resources/assets/skins/mainmenu.skin]: Agreed!

File: engine/src/main/resources/assets/ui/menu/selectGameScreenHost.ui [hunk 4]
```diff
@@ -0,0 +1,119 @@
+{
```
Comment (msteiger) [inline engine/src/main/resources/assets/ui/menu/selectGameScreenHost.ui]: Is this a duplicate?

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/484/
Hooray Jenkins reported success with all tests good!

**Comment by BdeBock** [comment]
Changed the name of subtitle2 to gameTypeTitle. Changed the static multi- and singleplayer strings to the translationSystem. Removed the duplicate selectGameScreenHost.ui.

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/485/
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/MainMenuScreen.java (hunk 1 above)
Comment (msteiger) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/MainMenuScreen.java]: The exact method call is: [NUIManager.pushScreen(String screenUri, Class<T> expectedType);](https://github.com/MovingBlocks/Terasology/blob/develop/engine/src/main/java/org/terasology/rendering/nui/NUIManager.java#L66). It returns a `SelectGameScreen` you can use directly.

**Comment by msteiger** [comment]
Thank you, Bas! Only one open issue to address and I will be happy to merge your contribution.

**Comment by BdeBock** [comment]
Team member is reviewing my solution for that issue, will be fixed tomorrow.

**Comment by BdeBock** [comment]
Last issue should be fixed now.

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/487/
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/488/
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/489/
Hooray Jenkins reported success with all tests good!

**Comment by msteiger** [comment]
Great - thank you for being so persistent.

**Comment by BdeBock** [comment]
No problem, thanks for the help.

**Comment by Cervator** [comment]
Much appreciated all :-)

Also I feel terrible saying this, but we have the new subtitle now on the single player and multiplayer screens - but if you then go to "Create" game in either case you are now again on an anonymous screen without the new subtitle... :D

Up for adding the subtitle on the create screen as well? On the plus side it should be a whole lot easier this time :-)

**Pull request merged**